From f0fdf9c36181858cc1546eec526b15018b509285 Mon Sep 17 00:00:00 2001
From: Omkar Anand Kulkarni <omkar.kulkarni@arm.com>
Date: Sat, 13 Apr 2024 21:27:24 +0530
Subject: [PATCH] feat(neoverse_rd): adds CPU error injection support

Neoverse CPU allows error injections of Corrected Error (CE), Deferred Error
(DE) and Uncontainable Error (UC) using the Pseudo Fault Generation
register that are implmented by the CPU ras extensions. Programming
these registers with valid values can generate error injections and help
in validating the error handling software.

Signed-off-by: Omkar Anand Kulkarni <omkar.kulkarni@arm.com>
Change-Id: I6c5e28bad6d91046ef6292bbfdb2ff4458e538b0

Signed-off-by: Jon Mason <jon.mason@arm.com>
Upstream-Status: Pending
---
 .../board/neoverse_rd/common/einj/nrd_einj.c  |  1 +
 .../neoverse_rd/common/include/nrd_einj.h     |  3 ++
 .../platform/rdn2/einj/nrd_einj_cpu.c         | 37 +++++++++++++++++++
 3 files changed, 41 insertions(+)
 create mode 100644 plat/arm/board/neoverse_rd/platform/rdn2/einj/nrd_einj_cpu.c

diff --git a/plat/arm/board/neoverse_rd/common/einj/nrd_einj.c b/plat/arm/board/neoverse_rd/common/einj/nrd_einj.c
index 65d798d29899..74e2f8fd8af9 100644
--- a/plat/arm/board/neoverse_rd/common/einj/nrd_einj.c
+++ b/plat/arm/board/neoverse_rd/common/einj/nrd_einj.c
@@ -89,6 +89,7 @@ static int err_inj_el3_interrupt_handler(uint32_t intr, uint32_t flags,
                        break;
                case EINJ_ERROR_PROCESSOR_UNCORRECTABLE_NONFATAL:
                        INFO("EINJ_ERROR_PROCESSOR_UNCORRECTABLE_NONFATAL\n");
+			nrd_cpu_inject_error();
                        break;
                case EINJ_ERROR_PROCESSOR_UNCORRECTABLE_FATAL:
                        INFO("EINJ_ERROR_PROCESSOR_UNCORRECTABLE_FATAL\n");
diff --git a/plat/arm/board/neoverse_rd/common/include/nrd_einj.h b/plat/arm/board/neoverse_rd/common/include/nrd_einj.h
index 1265c8f168ff..88f9a228ada0 100644
--- a/plat/arm/board/neoverse_rd/common/include/nrd_einj.h
+++ b/plat/arm/board/neoverse_rd/common/include/nrd_einj.h
@@ -52,3 +52,6 @@ int nrd_err_inj_init(void);
 
 /* Base element non-secure error injection helper */
 void nrd_sram_inject_error(void);
+
+/* CPU error injection helper */
+void nrd_cpu_inject_error(void);
diff --git a/plat/arm/board/neoverse_rd/platform/rdn2/einj/nrd_einj_cpu.c b/plat/arm/board/neoverse_rd/platform/rdn2/einj/nrd_einj_cpu.c
new file mode 100644
index 000000000000..178128ff104e
--- /dev/null
+++ b/plat/arm/board/neoverse_rd/platform/rdn2/einj/nrd_einj_cpu.c
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2022, ARM Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <common/debug.h>
+#include <drivers/delay_timer.h>
+#include <lib/extensions/ras_helpers.h>
+#include <plat/common/platform.h>
+
+void nrd_cpu_inject_error(void)
+{
+
+        INFO("cpu_id %d\n", plat_my_core_pos());
+
+        /* Select error record 1 which is for L1 L2 caches */
+        write_errselr_el1(0);
+
+        /* Enable CFI, FI and ED for node */
+        write_erxctlr_el1(0x109);
+
+        /* Clear PFG control register */
+        clear_cpu_pfg_ctrl_register();
+
+        /* Clear the status register */
+        write_erxstatus_el1(0xFFC00000);
+
+        /* set PFG countdown register to 1 */
+        write_cpu_pfg_cdn_register(1);
+
+	/* Injects CPU DE */
+        INFO("Injecting DE...\n");
+        write_cpu_pfg_ctrl_register_de();
+
+        INFO("ErrStatus = 0x%lx\n", read_erxstatus_el1());
+}
